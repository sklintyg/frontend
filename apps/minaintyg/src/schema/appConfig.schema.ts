import { z } from 'zod'

export const appConfigLinkSchema = z.object({
  key: z.string(),
  url: z.string(),
  text: z.string(),
  target: z.string(),
})

export const appConfigSchema = z.object({
  useMinifiedJavascript: z.boolean(),
  version: z.string(),
  buildNumber: z.string(),
  elva77MainUrl: z.string(),
  elva77LoginUrl: z.string(),
  applicationLogoutUrl: z.string(),
  miUserSurveyUrl: z.string(),
  miUserSurveyVersion: z.string(),
  miUserSurveyDateFrom: z.string(),
  miUserSurveyDateTo: z.string(),
  knownRecipients: z.array(z.object({ id: z.string(), name: z.string(), trusted: z.boolean() })),
  links: z.record(appConfigLinkSchema),
  banners: z.array(z.unknown()),
})

export type AppConfig = z.infer<typeof appConfigSchema>
export type AppConfigLink = z.infer<typeof appConfigLinkSchema>
