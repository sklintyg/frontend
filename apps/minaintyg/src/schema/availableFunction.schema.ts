import { z } from 'zod'

export const AvailableFunctionType = z.enum(['SEND_CERTIFICATE'])
export const AvailableFunctionInformationType = z.enum([''])

export const availableFunctionInformationSchema = z.object({
  type: AvailableFunctionInformationType,
  id: z.string(),
})

export const availableFunctionSchema = z.object({
  body: z.string().nullable(),
  description: z.string().nullable(),
  name: z.string().nullable(),
  title: z.string().nullable(),
  type: AvailableFunctionType,
  information: z.array(availableFunctionInformationSchema),
})

export type AvailableFunction = z.infer<typeof availableFunctionSchema>
