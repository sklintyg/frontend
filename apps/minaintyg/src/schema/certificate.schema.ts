import { z } from 'zod'

export const CertificateStatusEnum = z.enum(['NEW', 'REPLACED', 'SENT', 'NOT_SENT'])

export const certificateEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateTypeSchema = z.object({
  id: z.string(),
  version: z.string(),
  description: z.string().optional(),
})

export const certificateIssuerSchema = z.object({
  name: z.string(),
  phoneNumber: z.string(),
})

export const certificateUnitSchema = z.object({
  id: z.string(),
  name: z.string(),
  address: z.string(),
})

export const certificateMetadataSchema = z.object({
  issuer: certificateIssuerSchema,
  unit: certificateUnitSchema,
  name: z.string(),
  events: z.array(certificateEventSchema),
  statuses: z.array(CertificateStatusEnum),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
  summary: certificateSummarySchema,
})

export const certificateContentSchema = z.object({ heading: z.string(), body: z.string() })

export const certificateSchema = z.object({
  metadata: certificateMetadataSchema,
  content: z.array(certificateContentSchema),
})

export type CertificateContent = z.infer<typeof certificateContentSchema>
export type CertificateEvent = z.infer<typeof certificateEventSchema>
export type CertificateIssuer = z.infer<typeof certificateIssuerSchema>
export type CertificateMetadata = z.infer<typeof certificateMetadataSchema>
export type CertificateStatus = z.infer<typeof CertificateStatusEnum>
export type CertificateSummary = z.infer<typeof certificateSummarySchema>
export type CertificateUnit = z.infer<typeof certificateUnitSchema>
export type Certificate = z.infer<typeof certificateSchema>
