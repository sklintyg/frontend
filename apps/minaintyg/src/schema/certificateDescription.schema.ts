import { z } from 'zod'

export const certificateDescriptionSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string(),
  detailedDescription: z.string(),
  issuerTypeId: z.string(),
  cssPath: z.string(),
  scriptPath: z.string(),
  dependencyDefinitionPath: z.string(),
  defaultRecipient: z.string(),
  deprecated: z.boolean(),
  displayDeprecated: z.boolean(),
})

export type CertificateDescription = z.infer<typeof certificateDescriptionSchema>
