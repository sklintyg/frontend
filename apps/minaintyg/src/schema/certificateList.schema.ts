import { z } from 'zod'

export enum CertificateStatus {
  NEW = 'Nytt',
  REPLACED = 'Ers√§tter intyg',
  SENT = 'Skickat',
  NOT_SENT = 'Ej skickat',
}

export const certificateListSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateListEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateListIssuerSchema = z.object({ name: z.string() })
export const certificateUnitSchema = z.object({ id: z.string(), name: z.string() })
export const certificateTypeSchema = z.object({
  id: z.string(),
  name: z.string(),
  version: z.string(),
})

export const certificateListItemSchema = z.object({
  summary: certificateListSummarySchema,
  issuer: certificateListIssuerSchema,
  unit: certificateUnitSchema,
  events: z.array(certificateListEventSchema),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
})

export type CertificateListEvent = z.infer<typeof certificateListEventSchema>
export type CertificateListIssuer = z.infer<typeof certificateListIssuerSchema>
export type CertificateListUnit = z.infer<typeof certificateUnitSchema>
export type CertificateListSummary = z.infer<typeof certificateListSummarySchema>
export type CertificateListItem = z.infer<typeof certificateListItemSchema>
