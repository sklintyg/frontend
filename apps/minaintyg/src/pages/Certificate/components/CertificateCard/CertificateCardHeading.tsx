import { Link } from 'react-router-dom'
import type { CertificateStatus} from '../../../../schema/certificate.schema';
import { CertificateStatusEnum } from '../../../../schema/certificate.schema'
import { CertificateStatusBadge } from '../CertificateStatusBadge'

export function CertificateCardHeading({ title, id, statuses }: { title: string; id: string; statuses: CertificateStatus[] }) {
  return (
    <div className="mb-2.5 flex flex-col justify-between gap-2.5 md:flex-row">
      <h3 className="ids-heading-2 mb-0">
        <Link to={`/${id}`} className="ids-anchor">
          {title}
        </Link>
      </h3>
      <div className="flex gap-1 [&:not(:last-child)]:mb-2.5">
        {[...new Set(statuses)]
          .sort((status) => (status === CertificateStatusEnum.enum.NEW ? -1 : 1))
          .map((status) => (
            <CertificateStatusBadge key={status} status={status} />
          ))}
      </div>
    </div>
  )
}
