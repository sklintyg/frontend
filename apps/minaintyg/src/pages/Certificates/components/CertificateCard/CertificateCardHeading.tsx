import { Link } from 'react-router-dom'
import { CertificateStatus } from '../../../../schema/certificateList.schema'
import { CertificateStatusBadge } from '../CertificateStatusBadge'

export function CertificateCardHeading({ title, id, statuses }: { title: string; id: string; statuses: CertificateStatus[] }) {
  return (
    <div className="mb-2.5 flex flex-col justify-between gap-2.5 md:flex-row">
      <h3 className="ids-heading-2 mb-0">
        <Link to={`/intyg/${id}`} className="ids-anchor">
          {title}
        </Link>
      </h3>
      {[...new Set(statuses)].map((status) => (
        <CertificateStatusBadge key={status} status={status} />
      ))}
    </div>
  )
}
