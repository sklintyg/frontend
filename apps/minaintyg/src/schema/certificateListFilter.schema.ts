import { z } from 'zod'
import { CertificateStatus } from './certificateList.schema'

export const certificateFilterOptionsSchema = z.object({
  total: z.number(),
  statuses: z.array(z.nativeEnum(CertificateStatus)),
  units: z.array(z.string()),
  certificateTypes: z.array(z.string()),
  years: z.array(z.string()),
})

type SelectedOptions<T> = { [X in keyof T]: T[X] extends Array<unknown> ? T[X][number] : T[X] }

export type CertificateFilterOptions = z.infer<typeof certificateFilterOptionsSchema>
export type CertificateSelectedOptions = SelectedOptions<CertificateFilterOptions>
