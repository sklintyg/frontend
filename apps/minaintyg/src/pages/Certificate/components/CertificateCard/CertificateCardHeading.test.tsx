import { render } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { CertificateStatusEnum } from '../../../../schema/certificate.schema'
import { CertificateCardHeading } from './CertificateCardHeading'

function renderComponent() {
  return render(
    <MemoryRouter>
      <Routes>
        <Route
          path="/"
          element={
            <CertificateCardHeading title="Foo" id="123" statuses={[CertificateStatusEnum.enum.NOT_SENT, CertificateStatusEnum.enum.NEW]} />
          }
        />
      </Routes>
    </MemoryRouter>
  )
}

it('Should render correctly', () => {
  const { container } = renderComponent()
  expect(container).toMatchSnapshot()
})

it('Should sort status badges in correct order', () => {
  renderComponent()
  expect(Array.from(document.querySelectorAll('ids-badge')).map((el) => el.innerHTML)).toEqual(['Nytt', 'Ej skickat'])
})
