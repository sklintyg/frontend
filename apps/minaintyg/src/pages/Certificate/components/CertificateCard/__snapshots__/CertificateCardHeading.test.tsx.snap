// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render correctly 1`] = `
<div>
  <div
    class="mb-2.5 flex flex-col justify-between gap-2.5 md:flex-row"
  >
    <h3
      class="ids-heading-2 mb-0"
    >
      <a
        class="ids-anchor"
        href="/intyg/123"
      >
        Foo
      </a>
    </h3>
    <div
      class="flex gap-1 [&:not(:last-child)]:mb-2.5"
    >
      <ids-badge
        class="whitespace-nowrap"
      >
        Nytt
      </ids-badge>
      <ids-badge
        class="whitespace-nowrap"
      >
        Ej skickat
      </ids-badge>
    </div>
  </div>
</div>
`;
